### ISSUE TITLE
custom loss related to correlation with other model or feature

### ISSUE THREAD

**Issue by tarai2** [tarai2/Numer-ai#9]
#8 の最後:
・Featureに対してModel直交化=>パフォーマンス安定
(https://github.com/tarai2/Numer-ai/wiki/Model-Diagnostics:-Feature-Exposure)
・NNのtrainは特にlossの設計が大事
(https://github.com/tarai2/Numer-ai/wiki/NNをTournamentデータにフィットすることの難しさについて)
・2つのmodelをuncorrelatedに学習させる
(https://forum.numer.ai/t/learning-two-uncorrelated-models/400/16)
・torchを使ってxgbでカスタムlossを用いる方法
(https://forum.numer.ai/t/custom-loss-functions-for-xgboost-using-pytorch/960)

以上のポストから察するに、
uncorrelatedなモデル, FeatureExposureの低いモデルを複数作る => アンサンブルして強いモデルに
が正解なのではないか。この方針で行きたい。

**Comment by tarai2** [comment]
モデルに入れ込みたい要素メモ
- L1正則化
- corr (with feature)で罰則 ... featureExposureの低減  
- corr(with existing model's pred)で罰則 ... モデルの多様化  
- era wise boosting的な何か ... リターンの自己相関に対する罰則

**Comment by tarai2** [comment]
### custom loss for lgb  
fevalはscalarしか吐けない. 
- roundごとのFeatureExposures
- roundごとのera毎スコア  

のような多次元出力を保存しておきたい場合はcallbackが必要なので注意
